clang/lib/AST/PreorderAST.cpp

```diff
@@ -0,0 +1,202 @@
+//===------ PreorderAST.cpp: An n-ary preorder abstract syntax tree -------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements methods to create and manipulate an n-ary preorder
+//  abstract syntax tree which is used to semantically compare two expressions.
+//
+//===----------------------------------------------------------------------===//
+
+#include "clang/AST/PreorderAST.h"
+
+using namespace clang;
+
+bool PreorderAST::IsDeclOperand(Expr *E, DeclRefExpr *&D) {
+  if (auto *CE = dyn_cast_or_null<CastExpr>(E)) {
+    assert(CE->getSubExpr() && "invalid CastExpr expression");
+
+    if (CE->getCastKind() == CastKind::CK_LValueToRValue ||
+        CE->getCastKind() == CastKind::CK_ArrayToPointerDecay) {
+      E = Lex.IgnoreValuePreservingOperations(Ctx, CE->getSubExpr());
+      if (auto *DRE = dyn_cast_or_null<DeclRefExpr>(E)) {
+        D = DRE;
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+void PreorderAST::Create(ASTNode *N, Expr *E, ASTNode *Parent) {
+  if (!E)
+    return;
+
+  // When we invoke Create(N->Left, ...) or Create(N->Right, ...) we need to
+  // create the left or the right nodes with N as the parent node.
+  if (!N)
+    N = new ASTNode(Ctx, Parent);
+
+  // If the parent is non-null, make sure that the current node is marked as a
+  // child of the parent. As a convention, we create left children first.
+  if (Parent) {
+    if (!Parent->Left)
+      Parent->Left = N;
+    else
+      Parent->Right = N;
+  }
+
+  E = Lex.IgnoreValuePreservingOperations(Ctx, E);
+
+  // If E is a variable, store its name in the variable list for the current
+  // node. Initialize the count of the variable to 1.
+  DeclRefExpr *D;
+  if (IsDeclOperand(E, D)) {
```

Comment:
Could `IsDeclOperand` return a `DeclRefExpr *` rather than a boolean with a `DeclRefExpr *` out parameter? Then the caller could do `if (DeclRefExpr *D = IsDeclOperand(E))` (or potentially rename `IsDeclOperand` to `GetDeclOperand`).